src/gpt_j/attention.rs

```diff
@@ -0,0 +1,319 @@
+        let value = hidden_states.apply(&self.v_proj);
+
+        let mut query = Self::split_heads(&query, self.n_head, self.dim_per_head, true);
+        let mut key = Self::split_heads(&key, self.n_head, self.dim_per_head, true);
+        let mut value = Self::split_heads(&value, self.n_head, self.dim_per_head, false);
+
+        let mut seq_len = key.size()[1];
+        let mut offset = 0;
+
+        if let Some(layer_past) = layer_past {
+            offset = layer_past.prev_key.size()[layer_past.prev_key.size().len() - 2];
+            seq_len += offset
+        };
+
+        if let Some(rotary_dim) = self.rotary_dim {
+            let k_rot = key.slice(3, 0, rotary_dim, 1);
+            let k_pass = key.slice(3, rotary_dim, key.size()[3], 1);
+
+            let q_rot = query.slice(3, 0, rotary_dim, 1);
+            let q_pass = query.slice(3, rotary_dim, query.size()[3], 1);
+
+            let sincos = fixed_pos_embedding(&k_rot, seq_len);
+            let k_rot = apply_rotary_pos_emb(&k_rot, &sincos, offset);
+            let q_rot = apply_rotary_pos_emb(&q_rot, &sincos, offset);
+
+            key = Tensor::cat(&[k_rot, k_pass], -1);
+            query = Tensor::cat(&[q_rot, q_pass], -1);
+        } else {
+            let sincos = fixed_pos_embedding(&key, seq_len);
+            key = apply_rotary_pos_emb(&key, &sincos, offset);
+            query = apply_rotary_pos_emb(&query, &sincos, offset);
+        }
+
+        key = key.permute(&[0, 2, 1, 3]);
+        query = query.permute(&[0, 2, 1, 3]);
+
+        if let Some(layer_past) = layer_past {
+            key = Tensor::cat(&[&layer_past.prev_key, &key], -2);
+            value = Tensor::cat(&[&layer_past.prev_value, &value], -2);
+        }
+
+        let present = self.use_cache.then(|| LayerState {
+            prev_key: key.copy(),
+            prev_value: value.copy(),
+        });
+
+        let (attn_output, attn_weights) =
+            self.attention(&query, &key, &value, attention_mask, train);
+
+        let attn_output = Self::merge_heads(&attn_output, self.n_head, self.dim_per_head)
+            .apply(&self.out_proj)
+            .apply_t(&self.resid_dropout, train);
+
+        let attn_weights = self.output_attentions.then_some(attn_weights);
+
+        (attn_output, present, attn_weights)
+    }
+}
+
+fn fixed_pos_embedding(x: &Tensor, seq_len: i64) -> (Tensor, Tensor) {
+    let dim = x.size()[x.size().len() - 1];
+    let inv_freq = 1.0
+        / Tensor::pow_scalar(
+            10_000,
+            &(Tensor::arange_start_step(0, dim, 2, (x.kind(), x.device())) / dim),
+        );
+    let sinusoid_inp = Tensor::einsum(
+        "i , j -> i j",
+        &[Tensor::arange(seq_len, (x.kind(), x.device())), inv_freq],
+    );
+    (sinusoid_inp.sin(), sinusoid_inp.cos())
+}
+
+fn apply_rotary_pos_emb(x: &Tensor, (sin, cos): &(Tensor, Tensor), offset: i64) -> Tensor {
+    let sin = duplicate_interleave(sin).i((NewAxis, offset..x.size()[1] + offset, NewAxis, ..));
```

Comment:
Interesting first time I come across this

Reply:
I wasn't 100% confident it would work when I first wrote it, but it works !